### ISSUE TITLE
`ps.pgrep` causes backtrace for non-accessible processes on Windows

### ISSUE THREAD

**Issue by eliasp** [saltstack/salt#18689]
Affects Salt `2014.7.0` on Windows.

On Windows, some processes might not be accessible to the user running `salt-minion` such as system processes.

For example, the process `audiodg.exe` which is present on some Windows boxes is one of them and causes this backtrace, even when `pgrep`-ing for a completely different process:
```
$ salt MN1221-C0002 ps.pgrep firefox

MN1221-C0002:
    The minion function caused an exception: Traceback (most recent call last):
     File "salt/minion.py", line 1019, in _thread_return
      File "salt/modules/ps.py", line 292, in pgrep
      File "salt/modules/ps.py", line 67, in _get_proc_name
      File "psutil/__init__.py", line 490, in name
      File "psutil/_pswindows.py", line 206, in wrapper
      File "psutil/_pswindows.py", line 241, in name
      File "psutil/_pswindows.py", line 213, in wrapper
    AccessDenied: (pid=18904)
```
The first thing to do to fix this, is to handle the `psutil.NoSuchProcess` exception in the `psutil.process_iter()` iteration in `ps.pgrep`, but some more work in other places might be needed as well.

I'm working on a fix and will provide a PR once I have a proper solution.

**Comment by eliasp** [comment]
This seems to fix the issue here:

```diff
diff --git a/salt/modules/ps.py b/salt/modules/ps.py
index fa0fad2..26a1594 100644
--- a/salt/modules/ps.py
+++ b/salt/modules/ps.py
@@ -64,7 +64,18 @@ def _get_proc_name(proc):
 
     It's backward compatible with < 2.0 versions of psutil.
     '''
-    return proc.name() if PSUTIL2 else proc.name
+    ret = []
+    if PSUTIL2:
+        try:
+            ret = proc.name()
+        except (psutil.NoSuchProcess, psutil.AccessDenied):
+            pass
+    else:
+        try:
+            ret = proc.name
+        except (psutil.NoSuchProcess, psutil.AccessDenied):
+            pass
+    return ret
 
 
 def _get_proc_status(proc):
@@ -288,11 +299,14 @@ def pgrep(pattern, user=None, full=False):
 
     procs = []
     for proc in psutil.process_iter():
-        name_match = pattern in ' '.join(_get_proc_cmdline(proc)) if full \
-            else pattern in _get_proc_name(proc)
-        user_match = True if user is None else user == _get_proc_username(proc)
-        if name_match and user_match:
-            procs.append(_get_proc_pid(proc))
+        try:
+            name_match = pattern in ' '.join(_get_proc_cmdline(proc)) if full \
+                else pattern in _get_proc_name(proc)
+            user_match = True if user is None else user == _get_proc_username(proc)
+            if name_match and user_match:
+                procs.append(_get_proc_pid(proc))
+        except psutil.NoSuchProcess:
+            pass
     return procs or None
```

**Comment by eliasp** [comment]
Fix in #18689 got merged - closing.
